### ISSUE TITLE
TeamsUpgradePolicy: Apply config fails for global user policy in release 1.21.602.1

### ISSUE THREAD

**Issue by EricHindrichsen** [microsoft/Microsoft365DSC#1257]
#### Details of the scenario you tried and the problem that is occurring
We are running M365DSC within an Azure DevOps release pipeline. Our config is applied just fine with M365DSC version 1.21.526.2, but with version 1.21.602.1 the process throws the following error message:
```
Cannot bind parameter 'Filter' to the target. Exception setting "Filter": "Policy "Global" is not a user policy. You 
can assign only a user policy to a specific user."
    + CategoryInfo          : WriteError: (:) [], CimException
    + FullyQualifiedErrorId : ParameterBindingFailed,Microsoft.Rtc.Management.AD.Cmdlets.GetOcsUserCmdlet
    + PSComputerName        : localhost
The PowerShell DSC resource '[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc' with SourceInfo 
'DSC\Configuration\dev\config.dev.teams.ps1::88::9::TeamsUpgradePolicy' threw one 
or more non-terminating errors while running the Test-TargetResource functionality. These errors are logged to the ETW 
channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : localhost
```
#### Verbose logs showing the problem
```
2021-06-03T12:46:57.5511794Z VERBOSE: [fv-az182-664]: LCM:  [ Start  Resource ]  [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc]
2021-06-03T12:46:57.5515252Z VERBOSE: [fv-az182-664]: LCM:  [ Start  Test     ]  [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc]
2021-06-03T12:46:57.5521569Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] 
2021-06-03T12:46:57.5524962Z Importing the module MSFT_TeamsUpgradePolicy in force mode.
2021-06-03T12:46:57.5531246Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] Testing 
2021-06-03T12:46:57.5533804Z configuration of Team Upgrade Policy {Global}
2021-06-03T12:46:58.0977031Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] Checking
2021-06-03T12:46:58.0982431Z  the Teams Upgrade Policies
2021-06-03T12:46:59.1964015Z ##[error]Cannot bind parameter 'Filter' to the target. Exception setting "Filter": "Policy "Global" is not a user policy. You 
can assign only a user policy to a specific user."
    + CategoryInfo          : WriteError: (:) [], CimException
    + FullyQualifiedErrorId : ParameterBindingFailed,Microsoft.Rtc.Management.AD.Cmdlets.GetOcsUserCmdlet
    + PSComputerName        : localhost
2021-06-03T12:46:59.1983762Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] Found 
2021-06-03T12:46:59.1987459Z Teams Upgrade Policy with Identity {Global}
2021-06-03T12:46:59.1993977Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] Current 
2021-06-03T12:46:59.1997482Z Values: GlobalAdminAccount=System.Management.Automation.PSCredential; Identity=Global; MigrateMeetingsToTeams=False; 
2021-06-03T12:46:59.2003918Z Users=()
2021-06-03T12:46:59.2007601Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] Target 
2021-06-03T12:46:59.2014050Z Values: GlobalAdminAccount=System.Management.Automation.PSCredential; Identity=Global; MigrateMeetingsToTeams=False; 
2021-06-03T12:46:59.2017362Z Users=(); Verbose=True
2021-06-03T12:46:59.2023597Z VERBOSE: [fv-az182-664]:                            [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] 
2021-06-03T12:46:59.2028420Z Test-TargetResource returned True
2021-06-03T12:46:59.2034716Z VERBOSE: [fv-az182-664]: LCM:  [ End    Test     ]  [[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc] False in
2021-06-03T12:46:59.2038016Z  1.5940 seconds.
2021-06-03T12:46:59.2044164Z VERBOSE: [fv-az182-664]: LCM:  [ *FAILED*Compare  ]     Completed processing compare operation. The operation returned 
2021-06-03T12:46:59.2240136Z False.
2021-06-03T12:46:59.2255494Z ##[error]The PowerShell DSC resource '[TeamsUpgradePolicy]215c33c3-cdfa-4001-8a1f-bae517ff55bc' with SourceInfo 
'DSC\Configuration\dev\config.dev.teams.ps1::88::9::TeamsUpgradePolicy' threw one 
or more non-terminating errors while running the Test-TargetResource functionality. These errors are logged to the ETW 
channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : localhost
2021-06-03T12:46:59.2257883Z VERBOSE: Operation 'Invoke CimMethod' complete.
```
#### Suggested solution to the issue
I guess a change in the TeamsUpgradePolicy resource caused this issue.
#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
TeamsUpgradePolicy 215c33c3-cdfa-4001-8a1f-bae517ff55bc
        {
            GlobalAdminAccount     = $Credsglobaladmin;
            Identity               = "Global";
            MigrateMeetingsToTeams = $False;
            Users                  = @();
        }
```

#### The operating system the target node is running

Latest Azure DevOps release agent image
#### Version of the DSC module that was used ('dev' if using current dev branch)
1.21.602.1

**Comment by EricHindrichsen** [comment]
@NikCharlebois I found a hint online to try and not set the policy name (=settings are applied to default global policy), but this doesn't work in this case. I'm then getting an error message that the resource key can't be empty. Any chance to figure out which change in the resource module causes this issue?

**Comment by NikCharlebois** [comment]
Hi @EricHindrichsen,

My understanding is that you can't ever directly assign users to the global policy. Instead, settings for the Global policy will be assigned to users who are **not** assigned any other policies. Therefore, to "assign" the global policy really means not to assign any policies whatsoever to a given user. One option would be for the code to fail at compilation time if users are ever assigned to the Global policy in the config. The other option would be for the solution to automatically remove all policies for the users listed as being assigned to the Global policy, but then if you later assigned one, we would lose track of it.

e.g.

TeamsUpgradePolicy 53f95f28-9bb9-413b-ab72-22884063e3fe
        {
            GlobalAdminAccount     = $Credsglobaladmin;
            Identity               = "Global";
            MigrateMeetingsToTeams = $False;
            Users                  = @("Nik.Charlebois@$OrganizationName");
        }
        TeamsUpgradePolicy 404d09f5-1a3b-4dae-bc67-80858fd7f255
        {
            GlobalAdminAccount     = $Credsglobaladmin;
            Identity               = "UpgradeToTeams";
            MigrateMeetingsToTeams = $False;
            Users                  = @("Nik.Charlebois@$OrganizationName");
        }

The user would see all of it's policies unassigned with the Global block, only to be re-assigned the UpgradeToTeams policy right after. Thoughts?

**Comment by EricHindrichsen** [comment]
Hi @NikCharlebois,

thanks for the explanation! That makes sense. We have removed the "Users" config setting now for the global policy so it's no longer tracked. Since we are still tracking all remaining policies there won't be any undetected drifts. 

I also see that your latest release fixes the issue with Get-TargetResource, so thanks for that! However, with the new release we got an error message that version 1.5.0 of the module Microsoft.Graph.Authentication was not installed on our agent. Looks like the dependency on that is missing in the Microsoft 365DSC module definition.

But other than that this issue can be closed.
